### PR TITLE
 [Azure Search] SDK for Autocomplete

### PR DESCRIPTION
**Pull request by mhko**
@brjohnstmsft. 

While the swagger spec for autocomplete is being reviewed (https://github.com/Azure/azure-rest-api-specs-pr/pull/473). I wanted to share this PR with you for early visibility. Once the PR on the swagger spec is reviewed and merged, I will regenerate this code from the spec the right way and update this PR. 

## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] Please add REST spec PR link to the SDK PR
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by dsgouda** [comment]
@mhko Please fix test failures

**Comment by brjohnstmsft** [comment]
@mhko It looks like you forgot to add the new SessionRecords that you recorded for the Autocomplete tests. Always make sure that tests pass in Playback mode before submitting a PR.

**Comment by brjohnstmsft** [comment]
@mhko Please always include [Azure Search] in your commit messages

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/Models/AutocompleteParameters.cs [hunk 1]
```diff
@@ -16,44 +16,43 @@ namespace Microsoft.Azure.Search.Models
...
         /// highlights. Must be set with HighlightPreTag. If omitted, hit
-        /// highlighting of suggestions is disabled.</param>
+        /// highlighting is disabled.</param>
         /// <param name="highlightPreTag">A string tag that is prepended to hit
         /// highlights. Must be set with HighlightPostTag. If omitted, hit
-        /// highlighting of suggestions is disabled.</param>
+        /// highlighting is disabled.</param>
         /// <param name="minimumCoverage">A number between 0 and 100 indicating
-        /// the percentage of the index that must be covered by a suggestion
+        /// the percentage of the index that must be covered by am autocomplete
```
[the remark this answers is not in the document]
Comment (mhko) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/Models/AutocompleteParameters.cs]: Fixed in the swagger spec.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/Models/AutocompleteParameters.cs [hunk 2]
```diff
@@ -108,14 +107,14 @@ public AutocompleteParametersPayload(bool? fuzzy = default(bool?), string highli
 
         /// <summary>
         /// Gets or sets the comma-separated list of field names to consider
-        /// when querying for suggestions.
+        /// when querying for autucompleted terms.
```
[the remark this answers is not in the document]
Comment (mhko) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/Models/AutocompleteParameters.cs]: Thanks. Fixed in the swagger spec.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs [hunk 3]
```diff
@@ -65,46 +65,51 @@ internal DocumentsOperations(SearchIndexClient client)
...
         }
 
-        public async Task<AzureOperationResponse<AutocompleteResult>> AutocompleteGetWithHttpMessagesAsync(
+        public async Task<AzureOperationResponse<AutocompleteResult>> AutocompleteWithHttpMessagesAsync(
             AutocompleteMode autocompleteMode,
             string search,
             string suggesterName,
             SearchRequestOptions searchRequestOptions = default(SearchRequestOptions),
-            AutocompleteParametersPayload autocompleteParametersPayload = null,
+            AutocompleteParameters autocompleteParametersPayload = null,
```
[the remark this answers is not in the document]
Comment (mhko) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs]: Fixed.

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs [hunk 4]
```diff
@@ -85,23 +85,28 @@ internal DocumentsOperations(SearchIndexClient client)
...
                     searchRequestOptions,
-                    autocompleteParametersPayload,
+                    autocompleteParameters,
                     customHeaders,
                     cancellationToken);
             }
             else
             {
+                string searchFieldsStr = null;
+                if (autocompleteParameters != null && autocompleteParameters.SearchFields != null)
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs]: nit: You can shorten this to `if (autocompleteParameters?.SearchFields != null)`

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/DocumentsProxyOperations.cs [hunk 5]
```diff
@@ -409,7 +409,7 @@ internal DocumentsProxyOperations(SearchIndexClient client)
             }
             if (searchFields != null)
             {
-                _queryParameters.Add(string.Format("searchFields={0}", System.Uri.EscapeDataString(searchFields)));
+                _queryParameters.Add(string.Format("searchFields={0}", System.Uri.EscapeDataString(string.Join(",", searchFields))));
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Generated/DocumentsProxyOperations.cs]: Wow, AutoRest understands comma-separated parameters? That's pretty cool!

File: src/SDKs/Search/DataPlane/Search.Tests/SessionRecords/Microsoft.Azure.Search.Tests.GetAutocompleteTests/TestAutocompleteGetOneTerm.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
+{
+  "Entries": [
+    {
+      "RequestUri": "/subscriptions/3c729b2a-4f86-4bb2-abe8-4b8647af156c/resourceGroups/azsmnet4511/providers/Microsoft.Search/searchServices/azs-7938?api-version=2015-08-19",
+      "EncodedRequestUri": "L3N1YnNjcmlwdGlvbnMvM2M3MjliMmEtNGY4Ni00YmIyLWFiZTgtNGI4NjQ3YWYxNTZjL3Jlc291cmNlR3JvdXBzL2F6c21uZXQ0NTExL3Byb3ZpZGVycy9NaWNyb3NvZnQuU2VhcmNoL3NlYXJjaFNlcnZpY2VzL2F6cy03OTM4P2FwaS12ZXJzaW9uPTIwMTUtMDgtMTk=",
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Search.Tests/SessionRecords/Microsoft.Azure.Search.Tests.GetAutocompleteTests/TestAutocompleteGetOneTerm.json]: This looks very wrong. There's only a call to ARM to delete the test search service.

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/AutocompleteTests.cs [hunk 7]
```diff
@@ -172,7 +172,7 @@ protected void TestAutocompleteWithSelectedFields()
             SearchIndexClient client = GetClientForQuery();
             var autocompleteParameters = new AutocompleteParameters()
             {
-                SearchFields = "hotelName"
+                SearchFields = new[] { "hotelName" }
```
Comment (brjohnstmsft) [inline src/SDKs/Search/DataPlane/Search.Tests/Tests/AutocompleteTests.cs]: We need tests with more than one search field.

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs (hunk 4 above)
Comment (mhko) [reply src/SDKs/Search/DataPlane/Microsoft.Azure.Search.Data/Customizations/Documents/DocumentsOperations.cs]: Thanks!

File: src/SDKs/Search/DataPlane/Search.Tests/Tests/AutocompleteTests.cs (hunk 7 above)
Comment (mhko) [reply src/SDKs/Search/DataPlane/Search.Tests/Tests/AutocompleteTests.cs]: Added one more

File: src/SDKs/Search/DataPlane/Search.Tests/SessionRecords/Microsoft.Azure.Search.Tests.GetAutocompleteTests/TestAutocompleteGetOneTerm.json (hunk 6 above)
Comment (mhko) [reply src/SDKs/Search/DataPlane/Search.Tests/SessionRecords/Microsoft.Azure.Search.Tests.GetAutocompleteTests/TestAutocompleteGetOneTerm.json]: Thanks for noticing it. It looks like this session was recorded while refactoring was in progress. This test doesn't exist anymore.

**Review by bitsforthought — APPROVED** [review]

**Comment by brjohnstmsft** [comment]
@mhko This looks good. I'm going to go ahead and merge it, and you can submit another PR once the Swagger PR is merged and you're able to regenerate the code from the spec in Azure/master.
